packages/react-reconciler/src/ReactFiberHooks.js

```diff
@@ -170,6 +205,71 @@ function areHookInputsEqual(
   return true;
 }
 
+// till we have String::padEnd, a small function to
+// right-pad strings with spaces till a minimum length
+function padEndSpaces(string: string, length: number) {
+  if (string.length >= length) {
+    return string;
+  } else {
+    return string + ' ' + new Array(length - string.length).join(' ');
+  }
+}
+
+function flushHookMismatchWarnings() {
+  // we'll show the diff of the low level hooks,
+  // and a stack trace so the dev can locate where
+  // the first mismatch is coming from
+  if (__DEV__ && currentHookMismatch) {
+    const hookStackDiff = [];
+    let current = firstCurrentHook;
+    let previousOrder = [];
+    while (current) {
+      previousOrder.push(HookDevNames[((current: any): HookDev)._debugType]);
+      current = current.next;
+    }
+    let workInProgress = firstWorkInProgressHook;
+    let nextOrder = [];
+    while (workInProgress) {
+      nextOrder.push(HookDevNames[((workInProgress: any): HookDev)._debugType]);
+      workInProgress = workInProgress.next;
+    }
+    // some bookkeeping for formatting the output table
+    const columnLength = Math.max(
+      '  Previous render'.length,
+      ...previousOrder.map(hook => hook.length),
+    );
+    const hookStackHeader =
+      padEndSpaces('  Previous render', columnLength) +
+      '   Next render\n' +
+      padEndSpaces('  ---', columnLength) +
+      '   ---';
+    const hookStackLength = Math.max(previousOrder.length, nextOrder.length);
+    for (let i = 0; i < hookStackLength; i++) {
+      hookStackDiff.push(
+        (previousOrder[i] !== nextOrder[i] ? '> ' : '  ') +
+          padEndSpaces(previousOrder[i], columnLength) +
+          ' ' +
+          nextOrder[i],
+      );
+    }
+    warning(
+      false,
+      'React has detected a change in the order of hooks called by %s.\n' +
+        '%s\n' +
+        '%s\n\n' +
+        'Error trace:\n' +
+        '%s\n\n' +
+        'This will lead to bugs and errors if not fixed. ' +
+        'For more information, read the rules of hooks: https://fb.me/rules-of-hooks\n',
+      getComponentName(((currentlyRenderingFiber: any): Fiber).type),
```

Comment:
Nit: this is a bit hidden away. Let's read it very early so it's clearly visible this function assumes the variable is set.